emissionsapi/autoupdater.py

```diff
@@ -0,0 +1,167 @@
+import argparse
+import datetime
+import itertools
+import logging
+import multiprocessing
+import os
+import tempfile
+
+import sentinel5dl
+
+from emissionsapi import db, preprocess
+from emissionsapi.config import config
+
+# Logger
+logger = logging.getLogger(__name__)
+
+# Number of workers
+workers = config('workers') or 1
+
+# Earliest data to process
+earliest_data = datetime.datetime.fromisoformat(
+    config('earliest_data') or '2019-01-01')
+
+
+def generate_intervals(start, end, days=5):
+    '''Generator for intervals between start and end in days intervals.
+
+    :param start: Starting datetime for the intervals.
+    :type start: datetime.datetime
+    :param end: End datetime for the intervals.
+    :type end: datetime.datetime
+    :param days: Days for each interval, defaults to 5
+    :type days: int, optional
+    :yield: Start and end date of the next interval.
+    :rtype: tuple
+    '''
+    while start < end:
+        interval_end = min(start + datetime.timedelta(days=days), end)
+        yield start, interval_end
+        start = interval_end
+
+
+@db.with_session
+def get_intervals_to_process(session, table, exclude_existing=True):
+    '''Get all intervals with associated product files to process.
+    Intervals with existing data will be skipped.
+
+    :param session: SQLAlchemy Session
+    :type session: sqlalchemy.orm.session.Session
+    :param table: Table to search for existing data.
+    :type table: sqlalchemy.sql.schema.Table
+    :param ignore_existing: Exclude existing intervals, defaults to True
+    :type ignore_existing: bool, optional
+    :return: Generator yielding the next interval.
+    :rtype: generator
+    '''
+    # Ignore the existing data and return the full interval
+    if not exclude_existing:
+        return generate_intervals(earliest_data, datetime.datetime.now())
+
+    # No data present, so return the full interal
+    first, last = db.get_data_range(session, table).first()
+    if not first or not last:
+        return generate_intervals(earliest_data, datetime.datetime.now())
+
+    # Return interval excluding existing data
+    return itertools.chain(
+        generate_intervals(earliest_data, first),
+        generate_intervals(last, datetime.datetime.now())
```

Comment:
This wasn't a case I expected, since this tools was more about auto-magically filling the gaps between the data, but using `earliest_data` as a hard border makes sense to me.